portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java

```diff
@@ -0,0 +1,106 @@
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.appengine.repackaged.com.google.gson.Gson;
+import com.google.appengine.api.datastore.Query;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/files")
+public class FileHandlerServlet extends HttpServlet {
+
+private static final String CONTENT_TYPE_JSON = "application/json;";
+private static final String KEY_IMAGE_BLOBKEY = "blobKey";
+private static final String KEY_IMAGE_URL = "url";
+private static final String KIND_IMAGE = "image";
+private static final String PATH_HOME_PAGE = "/index.html";
+private final BlobstoreService blobService = BlobstoreServiceFactory.getBlobstoreService();
+private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+@Override
+protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+  List<Entity> files = datastore.prepare(new Query(KIND_IMAGE)).asList(FetchOptions.Builder.withDefaults());
+  String[] urls = files
+                    .stream()
+                    .map(e -> e.getProperty(KEY_IMAGE_URL))
+                    .toArray(String[]::new);
+  resp.setContentType(CONTENT_TYPE_JSON);
+  resp.getWriter().println(new Gson().toJson(urls));
+}
+
+ @Override
+ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+     Map<String, List<BlobKey>> blobs = blobService.getUploads(req);
+     List<BlobKey> blobKeys = blobs.get(KIND_IMAGE);
+
+     if (blobKeys == null || blobKeys.isEmpty()) {
+         // the user didn't upload a file, do nothing
+         resp.sendRedirect(PATH_HOME_PAGE);
+         return;
+     }
+
+     // The form only contains a single file input
+     BlobKey blobKey = blobKeys.get(0);
+
+     BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+     if (blobInfo.getSize() == 0) {
+         // the user didn't upload a file, cleanup
+         blobService.delete(blobKey);
+         resp.sendRedirect(PATH_HOME_PAGE);
+         return;
+     }
+     
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions opts = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    String imageUrl;
```

Comment:
I don't understand the logic here very well, why validate the url and get it's path, but on failure, just use it anyways? dropping everything except the path portion of the url seems like it could cause errors too?

Reply:
A lot of the code in this servlet comes from here: https://github.com/vedantroy/step-2020/blob/master/walkthroughs/week-4-libraries/blobstore/examples/hello-world-fetch/src/main/java/com/google/sps/servlets/FormHandlerServlet.java

The context is apparently I need to do some shenanigans so stuff works on both the dev and deployed server.